aten/src/ATen/native/sparse/SparseBlasImpl.cpp

```diff
@@ -212,6 +218,56 @@ Tensor& _compressed_row_strided_addmm_out(
 }
 
 namespace cpu {
+#if !AT_USE_MKL_SPARSE()
+template<typename scalar_t>
+void addmv_sparse_csr(
+        const scalar_t* mat_values,
+        const int64_t* crow_index,
+        const int64_t* col_index,
+        const int64_t rows,
+        const int64_t cols,
+        const scalar_t* vec,
+        const scalar_t alpha,
+        const scalar_t beta,
+        scalar_t* result) {
+        at::parallel_for(0, rows, 0, [&](int64_t rstart, int64_t rend) {
+          for(const auto row: c10::irange(rstart, rend)) {
+            scalar_t acc(0);
+            for(const auto idx: c10::irange(crow_index[row], crow_index[row+1])) {
+              acc += mat_values[idx]*vec[col_index[idx]];
+            }
+            result[row] = acc * alpha + result[row]*beta;
+          }
+        });
+}
+
+template<typename scalar_t>
+void addmv_sparse_bsr(
```

Comment:
You could turn this into a single function if you set blocksize to 1 for CSR.

Also note that blocksize for BSR is two values, so it can be blocksize (5, 10) for example.

Reply:
Added support for non-square blocks (though it's not supported by neither MKL nor accelerate)
Also, I would like to keep CSR separate, as it much faster than BSR, as it would contain superfluous divmod for every row and two extra multiplicatons for every column not to mention extra loop over the block size 1